shared/lib/src/bloc_complex/catalog/catalog_slice.dart

```diff
@@ -0,0 +1,43 @@
+import 'dart:math';
+
+import 'package:reactive_exploration/common/models/product.dart';
+import 'package:reactive_exploration/src/bloc_complex/catalog/catalog_page.dart';
+
+/// A slice of the catalog provided to an infinite-scrolling [ListView].
+///
+/// A [CatalogSlice] is a convenience class backed by an arbitrary
+/// number of [CatalogPage]s. Most of the time, there is more than one
+/// [CatalogPage] in memory — but the view code should _not_ need to worry about
+/// that.
+class CatalogSlice {
+  final List<CatalogPage> _pages;
+
+  /// The index at which this slice starts to provide [Product]s.
+  final int startIndex;
+
+  /// Whether or not this slice is the end of the catalog.
+  ///
+  /// Currently always `true` as our catalog is infinite.
+  final bool hasNext;
+
+  const CatalogSlice(this._pages, this.startIndex, this.hasNext);
```

Comment:
Is there a reason why you're not calculating startIndex from _pages inside the constructor rather than passing it as a param?